src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -81,9 +81,18 @@ ErrorMessageOr<uint64_t> ExecuteOrDie(pid_t pid, uint64_t address_code, uint64_t
   RegisterState original_registers;
   OUTCOME_TRY(original_registers.BackupRegisters(pid));
 
-  RegisterState registers_set_rip = original_registers;
-  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
-  OUTCOME_TRY(registers_set_rip.RestoreRegisters());
+  RegisterState registers_for_execution = original_registers;
+  registers_for_execution.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  // The calling convention for x64 assumes the 128 bytes below rsp to be usable as a scratch pad
+  // for the current function. This area is called the 'red zone'. The function we interrupted might
+  // have stored temporary data in the red zone and the function we are about to execute might do
+  // the same. To keep them separated we decrement rsp by 128 bytes.
+  // The calling convention also asks for rsp being a multiple of 16 so we additionally round down
+  // to the next multiople of 16.
+  const uint64_t old_rsp = original_registers.GetGeneralPurposeRegisters()->x86_64.rsp;
+  const uint64_t aligned_rsp_below_red_zone = (old_rsp - 128) / 16 * 16;
+  registers_for_execution.GetGeneralPurposeRegisters()->x86_64.rsp = aligned_rsp_below_red_zone;
```

Comment:
I think doing the offset here is dangerous as any injected code that pushes to the stack will mess with the alignment of rsp and with the red zone. Should we instead do this right before the call instruction in the injected code?

Reply:
Actually, I read a bit more about the red zone, it's not at all similar to the shadow space on Windows. If I understand correctly, we don't need to change RSP at all, the red zone is simply a guarantee that the 128 bytes below RSP "will not be asynchronously clobbered by signals or interrupt handlers, and thus can use it for scratch data, without explicitly moving the stack pointer." (https://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64#the-red-zone). 

We shouldn't modify RSP other than by making sure it is 16 bytes aligned.

Reply:
I found that same document. It is linked in the bug, maybe I should add the link here.

>  it's not at all similar to the shadow space on Windows

Well, it is space on the stack usable for temporary variables without changing rsp. On windows it is above rsp and one can rely on the caller to allocate it. Here the area below rsp is implicitly allocated. 

>will not be asynchronously clobbered ...

But this is what we might do: Assume that the function we interrupt writes to the red zone (it might be a leaf function). The function we call might do the same. When we continue execution of the interrupted function the red zone will contain rubbish.
To avoid that I move rsp below the current red zone to a untouched part of the stack where the calling convention does not make any guarantees. When the regular execution continues rsp is restored and the interrupted function gets back its unmodified red zone.

Reply:
> will not be asynchronously clobbered by signals or interrupt handlers

If I understand correctly the point here is exactly that we are "asynchronously clobbering" it, hence breaking the calling convention if we don't change rsp.

Reply:
Yes, never mind, it makes sense to modify RSP before executing the injected code precisely because we don't want to clobber the existing red zone. I wonder if we are moving it enough then. The red zone is 128 bytes after the local variables. Could it be that a function has used the red zone but not yet incremented RSP for its local variables (which we could have more than 128 bytes of? In that case calling our injected code could clobber the red zone.

Reply:
> Could it be that a function has used the red zone but not yet decremented RSP for its local variables (which we could have more than 128 bytes of)?

Is that a problem? The way I understand it the red zone is defined based on the current value of RSP in every moment, so if a function uses more than 128 bytes past RSP before initially decrementing RSP then the function is already breaking the calling convention.

Reply:
Ah yes, this makes more sense now, thanks for the input.

Reply:
Okay, yes that also makes sense. 

As commented under "conversations": Section 3.2.2 here: https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.99.pdf reads as if the stack frame is set up before anything else happens. But it could be made clearer there.